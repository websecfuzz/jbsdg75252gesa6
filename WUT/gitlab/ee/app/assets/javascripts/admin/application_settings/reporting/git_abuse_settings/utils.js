import { parseBoolean } from '~/lib/utils/common_utils';

export const parseFormProps = ({
  maxNumberOfRepositoryDownloads,
  maxNumberOfRepositoryDownloadsWithinTimePeriod,
  gitRateLimitUsersAllowlist,
  gitRateLimitUsersAlertlist,
  autoBanUserOnExcessiveProjectsDownload,
}) => ({
  maxNumberOfRepositoryDownloads: parseInt(maxNumberOfRepositoryDownloads, 10),
  maxNumberOfRepositoryDownloadsWithinTimePeriod: parseInt(
    maxNumberOfRepositoryDownloadsWithinTimePeriod,
    10,
  ),
  gitRateLimitUsersAllowlist: JSON.parse(gitRateLimitUsersAllowlist),
  gitRateLimitUsersAlertlist: JSON.parse(gitRateLimitUsersAlertlist),
  autoBanUserOnExcessiveProjectsDownload: parseBoolean(autoBanUserOnExcessiveProjectsDownload),
});
