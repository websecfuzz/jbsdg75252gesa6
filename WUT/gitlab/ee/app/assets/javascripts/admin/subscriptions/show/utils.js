import {
  subscriptionTypes,
  offlineCloudLicenseText,
  onlineCloudLicenseText,
  licenseFileText,
} from './constants';

export const getLicenseFromData = ({ data } = {}) => data?.gitlabSubscriptionActivate?.license;
export const getErrorsAsData = ({ data } = {}) => data?.gitlabSubscriptionActivate?.errors || [];

export function getLicenseTypeLabel(type) {
  switch (type) {
    case subscriptionTypes.OFFLINE_CLOUD:
      return offlineCloudLicenseText;
    case subscriptionTypes.ONLINE_CLOUD:
      return onlineCloudLicenseText;
    default:
      return licenseFileText;
  }
}
