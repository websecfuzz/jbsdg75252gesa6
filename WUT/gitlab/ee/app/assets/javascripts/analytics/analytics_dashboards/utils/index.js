import produce from 'immer';
import { SESSIONS_TABLE_NAME } from 'ee/analytics/analytics_dashboards/constants';
import { DATE_RANGE_FILTER_DIMENSIONS } from 'ee/analytics/analytics_dashboards/data_sources/cube_analytics';
import getAllCustomizableDashboardsQuery from '../graphql/queries/get_all_customizable_dashboards.query.graphql';
import getCustomizableDashboardQuery from '../graphql/queries/get_customizable_dashboard.query.graphql';
import { TYPENAME_ANALYTICS_DASHBOARD_PANEL } from '../graphql/constants';

/**
 * Given a CubeJS property (e.g. `Sessions.count`), get the schema name (e.g. `Sessions`).
 */
export function getMetricSchema(metric) {
  return metric?.split('.')[0];
}

/**
 * Filters an array of dimensions by schema
 */
export function getDimensionsForSchema(selectedSchema, availableDimensions) {
  if (!selectedSchema) return [];

  return availableDimensions.filter(({ name }) => getMetricSchema(name) === selectedSchema);
}

/**
 * Selects a time dimension for a given schema
 */
export function getTimeDimensionForSchema(selectedSchema, availableTimeDimensions) {
  if (!selectedSchema) return null;

  const timeDimensions = availableTimeDimensions.filter(
    ({ name }) => getMetricSchema(name) === selectedSchema,
  );

  if (timeDimensions.length === 1) {
    // We only allow filtering by a single timeDimension. We expect most of our cubes to only have a single time dimension.
    return timeDimensions.at(0);
  }

  if (selectedSchema === SESSIONS_TABLE_NAME) {
    // Our `Sessions` cube is different, having both `startsAt`, `endsAt` timeDimensions.
    // We want to explicitly select the right dimension for Sessions so have this hardcoded lookup
    const sessionsDimensionName = DATE_RANGE_FILTER_DIMENSIONS[SESSIONS_TABLE_NAME.toLowerCase()];
    return timeDimensions.find(({ name }) => name === sessionsDimensionName);
  }

  // An unknown situation where a cube other than Sessions has more than one timeDimension. Hide it from the UI.
  return null;
}

/**
 * Updates a dashboard detail in cache from getProductAnalyticsDashboard:{slug}
 */
const updateDashboardDetailsApolloCache = ({
  apolloClient,
  dashboard,
  slug,
  fullPath,
  isProject,
  isGroup,
}) => {
  const getDashboardDetailsQuery = {
    query: getCustomizableDashboardQuery,
    variables: {
      fullPath,
      slug,
      isProject,
      isGroup,
    },
  };
  const sourceData = apolloClient.readQuery(getDashboardDetailsQuery);
  if (!sourceData) {
    // Dashboard details not yet in cache, must be a new dashboard, nothing to update
    return;
  }

  const data = produce(sourceData, (draftState) => {
    const { nodes } = isProject
      ? draftState.project.customizableDashboards
      : draftState.group.customizableDashboards;
    const updateIndex = nodes.findIndex((node) => node.slug === slug);

    if (updateIndex < 0) return;

    const updateNode = nodes[updateIndex];

    nodes.splice(updateIndex, 1, {
      ...updateNode,
      ...dashboard,
      panels: {
        ...updateNode.panels,
        nodes:
          dashboard.panels?.map((panel) => {
            const { id, ...panelRest } = panel;
            return { __typename: TYPENAME_ANALYTICS_DASHBOARD_PANEL, ...panelRest };
          }) || [],
      },
    });
  });

  apolloClient.writeQuery({
    ...getDashboardDetailsQuery,
    data,
  });
};

/**
 * Adds/updates a newly created dashboard to the dashboards list cache from getAllCustomizableDashboardsQuery
 */
const updateDashboardsListApolloCache = ({
  apolloClient,
  dashboardSlug,
  dashboard,
  fullPath,
  isProject,
  isGroup,
}) => {
  const getDashboardListQuery = {
    query: getAllCustomizableDashboardsQuery,
    variables: {
      fullPath,
      isProject,
      isGroup,
    },
  };
  const sourceData = apolloClient.readQuery(getDashboardListQuery);
  if (!sourceData) {
    // Dashboard list not yet loaded in cache, nothing to update
    return;
  }

  const data = produce(sourceData, (draftState) => {
    const { panels, ...dashboardWithoutPanels } = dashboard;
    const { nodes } = isProject
      ? draftState.project.customizableDashboards
      : draftState.group.customizableDashboards;

    const updateIndex = nodes.findIndex(({ slug }) => slug === dashboardSlug);

    // Add new dashboard if it doesn't exist
    if (updateIndex < 0) {
      nodes.push(dashboardWithoutPanels);
      return;
    }

    nodes.splice(updateIndex, 1, {
      ...nodes[updateIndex],
      ...dashboardWithoutPanels,
    });
  });

  apolloClient.writeQuery({
    ...getDashboardListQuery,
    data,
  });
};

export const updateApolloCache = ({
  apolloClient,
  slug,
  dashboard,
  fullPath,
  isProject,
  isGroup,
}) => {
  // TODO: modify to support removing dashboards from cache https://gitlab.com/gitlab-org/gitlab/-/issues/425513
  updateDashboardDetailsApolloCache({
    apolloClient,
    dashboard,
    slug,
    fullPath,
    isProject,
    isGroup,
  });
  updateDashboardsListApolloCache({ apolloClient, slug, dashboard, fullPath, isProject, isGroup });
};
