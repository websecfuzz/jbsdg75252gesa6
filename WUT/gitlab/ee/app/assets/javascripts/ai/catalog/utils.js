import { formValidators } from '@gitlab/ui/dist/utils';
import { s__, sprintf } from '~/locale';

export const createFieldValidators = ({ requiredLabel, maxLength } = {}) => {
  const validators = [];

  if (requiredLabel !== undefined) {
    validators.push(formValidators.required(requiredLabel));
  }

  if (maxLength !== undefined) {
    validators.push(
      formValidators.factory(
        sprintf(
          s__('AICatalog|Input cannot exceed %{value} characters. Please shorten your input.'),
          {
            value: maxLength,
          },
        ),
        (value) => (value?.length || 0) <= maxLength,
      ),
    );
  }

  return validators;
};
