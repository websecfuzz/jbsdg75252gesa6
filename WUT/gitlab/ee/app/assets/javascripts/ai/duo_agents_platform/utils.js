import { s__ } from '~/locale';
import { humanize } from '~/lib/utils/text_utility';

export const formatAgentDefinition = (agentDefinition) => {
  return humanize(agentDefinition || s__('DuoAgentsPlatform|Agent session'));
};

export const formatAgentFlowName = (agentDefinition, id) => {
  return `${formatAgentDefinition(agentDefinition)} #${id}`;
};

export const formatAgentStatus = (status) => {
  return status ? humanize(status.toLowerCase()) : s__('DuoAgentsPlatform|Unknown');
};
