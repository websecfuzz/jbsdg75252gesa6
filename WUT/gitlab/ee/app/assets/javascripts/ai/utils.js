import { duoChatGlobalState } from '~/super_sidebar/constants';
import { setCookie } from '~/lib/utils/common_utils';
import {
  DUO_AGENTIC_MODE_COOKIE,
  DUO_AGENTIC_MODE_COOKIE_EXPIRATION,
} from 'ee/ai/tanuki_bot/constants';

export const concatStreamedChunks = (arr) => {
  if (!arr) return '';

  let end = arr.findIndex((el) => !el);

  if (end < 0) end = arr.length;

  return arr.slice(0, end).join('');
};

/**
 * setCookie wrapper with duo agentic mode constances.
 *
 * @param {isAgenticMode} Boolean - Value to save
 * @returns {void}
 */
export const saveDuoAgenticModePreference = (isAgenticMode) => {
  setCookie(DUO_AGENTIC_MODE_COOKIE, isAgenticMode, {
    expires: DUO_AGENTIC_MODE_COOKIE_EXPIRATION,
  });
};

/**
 * Swith duo chat based on agenticMode value and save to cookie based on
 * saveCookie value.
 *
 * @param {agenticMode} Boolean - Agentic mode state
 * @param {saveCookie} Boolean - Save to cookie flag
 * @returns {void}
 */

export const setAgenticMode = (agenticMode = true, saveCookie = false) => {
  duoChatGlobalState.isShown = !agenticMode;
  duoChatGlobalState.isAgenticChatShown = agenticMode;

  if (saveCookie) {
    saveDuoAgenticModePreference(agenticMode);
  }
};

/**
 * Sends a command to DuoChat to execute on. This should be use for
 * a single command.
 *
 * @param {question} String - Prompt to send to the chat endpoint
 * @param {resourceId} String - Unique ID to bind the streaming
 * @param {variables} Object - Additional variables to pass to graphql chat mutation
 */
export const sendDuoChatCommand = ({ question, resourceId, variables = {} } = {}) => {
  if (!question || !resourceId) {
    throw new Error('Both arguments `question` and `resourceId` are required');
  }
  setAgenticMode(false, true);

  window.requestIdleCallback(() => {
    duoChatGlobalState.commands.push({
      question,
      resourceId,
      variables,
    });
  });
};

export const clearDuoChatCommands = () => {
  duoChatGlobalState.commands = [];
};

/**
 * Converts a text string into a URL-friendly format for event tracking.
 *
 * - Converts to lowercase
 * - Removes special characters
 * - Replaces spaces with underscores
 * - Limits length to 50 characters
 *
 * @param {string} text - The text to convert
 * @returns {string} The formatted event label
 */
export const generateEventLabelFromText = (text) => {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9\s]/g, '')
    .replace(/\s+/g, '_')
    .substring(0, 50);
};

export const utils = {
  concatStreamedChunks,
  generateEventLabelFromText,
};
