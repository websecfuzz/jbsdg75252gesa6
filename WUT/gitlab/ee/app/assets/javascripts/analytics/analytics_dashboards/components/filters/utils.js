import { queryToObject } from '~/lib/utils/url_utility';
import { formatDate, newDate } from '~/lib/utils/datetime_utility';
import { ISO_SHORT_FORMAT } from '~/vue_shared/constants';
import {
  convertObjectPropsToCamelCase,
  convertObjectPropsToSnakeCase,
  parseBoolean,
} from '~/lib/utils/common_utils';
import {
  filterToQueryObject,
  urlQueryToFilter,
} from '~/vue_shared/components/filtered_search_bar/filtered_search_utils';
import {
  START_DATES,
  DATE_RANGE_OPTIONS,
  DATE_RANGE_OPTION_CUSTOM,
  DATE_RANGE_OPTION_KEYS,
  DEFAULT_SELECTED_DATE_RANGE_OPTION,
  FILTERED_SEARCH_SUPPORTED_TOKENS,
  PROJECT_FILTER_QUERY_NAME,
} from './constants';

const isCustomOption = (option) => option && option === DATE_RANGE_OPTION_CUSTOM;

export const getDateRangeOption = (optionKey) => DATE_RANGE_OPTIONS[optionKey] || null;

export const dateRangeOptionToFilter = ({ startDate, endDate, key }) => ({
  startDate,
  endDate,
  dateRangeOption: key,
});

const DEFAULT_FILTER = dateRangeOptionToFilter(
  DATE_RANGE_OPTIONS[DEFAULT_SELECTED_DATE_RANGE_OPTION],
);

export const buildDefaultDashboardFilters = (queryString, dashboardDefaultFilters = {}) => {
  const {
    dateRangeOption,
    startDate,
    endDate,
    filterAnonUsers,
    [PROJECT_FILTER_QUERY_NAME]: projectFullPath,
  } = convertObjectPropsToCamelCase(queryToObject(queryString, { gatherArrays: true }));
  const searchFilters = urlQueryToFilter(queryString, {
    filterNamesAllowList: FILTERED_SEARCH_SUPPORTED_TOKENS,
  });

  const isDateRangeFilterEnabled = dashboardDefaultFilters?.dateRange?.enabled;
  const optionKey = dateRangeOption || dashboardDefaultFilters?.dateRange?.defaultOption;
  const dateRangeOverride = DATE_RANGE_OPTION_KEYS.includes(optionKey)
    ? dateRangeOptionToFilter(getDateRangeOption(optionKey))
    : {};

  const customDateRange = isCustomOption(optionKey);

  return {
    ...(isDateRangeFilterEnabled && {
      ...DEFAULT_FILTER,
      // Override default filter with user defined option
      ...dateRangeOverride,
      // Override date range when selected option is custom date range
      ...(customDateRange && { startDate: newDate(startDate) }),
      ...(customDateRange && { endDate: newDate(endDate) }),
    }),
    filterAnonUsers: parseBoolean(filterAnonUsers),
    searchFilters,
    projectFullPath: projectFullPath || null,
  };
};

export const filtersToQueryParams = ({
  dateRangeOption,
  startDate,
  endDate,
  filterAnonUsers,
  searchFilters,
  projectFullPath,
}) => {
  const customDateRange = isCustomOption(dateRangeOption);

  const searchFiltersQueryObj = filterToQueryObject(searchFilters);
  const additionalFiltersQueryObj = convertObjectPropsToSnakeCase({
    dateRangeOption,
    // Clear the date range unless the custom date range is selected
    startDate: customDateRange ? formatDate(startDate, ISO_SHORT_FORMAT) : null,
    endDate: customDateRange ? formatDate(endDate, ISO_SHORT_FORMAT) : null,
    // Clear the anon users filter unless truthy
    filterAnonUsers: filterAnonUsers || null,
    [PROJECT_FILTER_QUERY_NAME]: projectFullPath || null,
  });

  return { ...additionalFiltersQueryObj, ...searchFiltersQueryObj };
};

export function isDashboardFilterEnabled(filter) {
  return filter?.enabled || false;
}

export const getStartDate = (option) =>
  START_DATES[option] ?? START_DATES[DEFAULT_SELECTED_DATE_RANGE_OPTION];
