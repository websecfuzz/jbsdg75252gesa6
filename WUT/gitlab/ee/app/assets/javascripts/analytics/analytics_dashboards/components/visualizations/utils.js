import { isNil } from 'lodash';
import { isNumeric } from '~/lib/utils/number_utils';
import { formatNumber, n__, __, sprintf } from '~/locale';
import { formatDate, humanizeTimeInterval } from '~/lib/utils/datetime/date_format_utility';
import { formatAsPercentage } from 'ee/analytics/dora/components/util';
import { NULL_SERIES_ID } from 'ee/analytics/shared/constants';
import { UNITS } from '~/analytics/shared/constants';

function isIsoDateString(dateString) {
  // Matches an ISO date string in the format `2024-03-14T00:00:00.000`
  const isoDateRegex = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}$/;
  return isoDateRegex.test(dateString);
}

export function formatVisualizationValue(value) {
  if (isIsoDateString(value)) {
    return formatDate(value);
  }

  if (isNumeric(value)) {
    return formatNumber(parseInt(value, 10));
  }

  return value;
}

export function formatVisualizationTooltipTitle(title, params) {
  const value = params?.seriesData?.at(0)?.value?.at(0);

  if (isIsoDateString(value)) {
    const formattedDate = formatDate(value);
    return title.replace(value, formattedDate);
  }

  return title;
}

export function customFormatVisualizationTooltipTitle(params, formatter) {
  const xAxisValue = params?.seriesData?.at(0)?.value?.at(0);

  if (isNil(xAxisValue)) return '';

  return formatter(xAxisValue);
}

export const humanizeDisplayUnit = ({ unit, data = 0 }) => {
  switch (unit) {
    case 'days':
      return n__('day', 'days', data === '-' ? 0 : data);
    case 'per_day':
      return __('/day');
    case 'percent':
      return '%';
    default:
      return unit;
  }
};

/**
 * Humanizes values to be displayed in chart tooltips
 *
 * @param {string} unit â€“ The unit of measurement to be used for metric
 * @param {number} value - The value of the metric
 * @returns {string|number} - Humanized tooltip value
 */
export const humanizeChartTooltipValue = ({ unit, value } = {}) => {
  if (isNil(value)) return __('No data');

  switch (unit) {
    case UNITS.DAYS:
      return n__('%d day', '%d days', value);
    case UNITS.PER_DAY:
      return sprintf(__('%{value} /day'), { value });
    case UNITS.PERCENT:
      return formatAsPercentage(value);
    case UNITS.TIME_INTERVAL:
      return humanizeTimeInterval(value);
    default:
      return value;
  }
};

export const calculateDecimalPlaces = ({ data, decimalPlaces } = {}) => {
  return (data && parseInt(decimalPlaces, 10)) || 0;
};

export const removeNullSeries = (seriesData) => {
  return seriesData?.filter(({ seriesId }) => seriesId !== NULL_SERIES_ID);
};
