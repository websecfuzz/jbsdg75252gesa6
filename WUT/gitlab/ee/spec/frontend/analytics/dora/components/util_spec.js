import lastWeekData from 'test_fixtures/api/dora/metrics/daily_lead_time_for_changes_for_last_week.json';
import deploymentFrequencyLastWeekData from 'test_fixtures/api/dora/metrics/daily_deployment_frequency_for_last_week.json';
import { secondsToDays } from '~/analytics/shared/utils';
import { buildNullSeries } from 'ee/analytics/shared/utils';
import {
  apiDataToChartSeries,
  seriesToAverageSeries,
  seriesToMedianSeries,
  extractTimeSeriesTooltip,
  formatAsPercentage,
  forecastDataToSeries,
  extractOverviewMetricsQueryParameters,
} from 'ee/analytics/dora/components/util';
import { forecastDataToChartDate } from './helpers';
import { mockLastWeekData, mockLastWeekRawForecastData } from './mock_data';

const NO_DATA_MESSAGE = 'No data available';

describe('ee/analytics/dora/components/util.js', () => {
  describe('apiDataToChartSeries', () => {
    it('transforms the data from the API into data the chart component can use', () => {
      const apiData = [
        // This is the date format we expect from the API
        { value: 5, date: '2015-06-28' },

        // But we should support _any_ date format
        { value: 1, date: '2015-06-28T20:00:00.000-0400' },
        { value: 8, date: '2015-07-01T00:00:00.000Z' },
      ];

      const startDate = new Date(2015, 5, 26, 10);
      const endDate = new Date(2015, 6, 4, 10);
      const chartTitle = 'Chart title';

      const expected = [
        {
          name: chartTitle,
          data: [
            ['Jun 26, 2015', 0],
            ['Jun 27, 2015', 0],
            ['Jun 28, 2015', 5],
            ['Jun 29, 2015', 1],
            ['Jun 30, 2015', 0],
            ['Jul 01, 2015', 8],
            ['Jul 02, 2015', 0],
            ['Jul 03, 2015', 0],
          ],
        },
      ];

      expect(apiDataToChartSeries(apiData, startDate, endDate, chartTitle)).toEqual(expected);
    });
  });

  describe('lead time data', () => {
    it('returns the correct lead time chart data after all processing of the API response', () => {
      const chartData = buildNullSeries({
        seriesData: apiDataToChartSeries(
          lastWeekData,
          new Date(2015, 5, 27, 10),
          new Date(2015, 6, 4, 10),
          'Lead time',
          null,
        ),
        nullSeriesTitle: NO_DATA_MESSAGE,
      });

      expect(chartData).toMatchSnapshot();
    });
  });

  describe('seriesToAverageSeries', () => {
    const seriesName = 'Average';

    it('returns an empty object if chart data is undefined', () => {
      const data = seriesToAverageSeries(undefined, seriesName);

      expect(data).toStrictEqual({});
    });

    it('returns an empty object if chart data is blank', () => {
      const data = seriesToAverageSeries(null, seriesName);

      expect(data).toStrictEqual({});
    });

    it('returns the correct average values', () => {
      const data = seriesToAverageSeries(
        [
          ['Jul 01, 2015', 2],
          ['Jul 02, 2015', 3],
          ['Jul 03, 2015', 4],
        ],
        seriesName,
      );

      expect(data).toStrictEqual({
        name: seriesName,
        data: [
          ['Jul 01, 2015', 3],
          ['Jul 02, 2015', 3],
          ['Jul 03, 2015', 3],
        ],
      });
    });
  });

  describe('seriesToMedianSeries', () => {
    const seriesName = 'Median';

    it('returns an empty object if chart data is undefined', () => {
      const data = seriesToMedianSeries(undefined, seriesName);

      expect(data).toStrictEqual({});
    });

    it('returns an empty object if chart data is blank', () => {
      const data = seriesToMedianSeries(null, seriesName);

      expect(data).toStrictEqual({});
    });

    it('returns the correct median values', () => {
      const data = seriesToMedianSeries(
        [
          ['Jul 01, 2015', 1],
          ['Jul 02, 2015', 3],
          ['Jul 03, 2015', 10],
        ],
        seriesName,
      );

      expect(data).toStrictEqual({
        name: seriesName,
        data: [
          ['Jul 01, 2015', 3],
          ['Jul 02, 2015', 3],
          ['Jul 03, 2015', 3],
        ],
      });
    });
  });

  describe('extractTimeSeriesTooltip', () => {
    const fakeChartTitle = 'cool-chart-title';
    const params = { seriesData: [{ data: ['Apr 7', 5328] }, { data: ['Apr 7', 4000] }, {}] };

    it('displays a humanized version of the time interval in the tooltip', () => {
      const { tooltipValue } = extractTimeSeriesTooltip(params, fakeChartTitle);

      expect(tooltipValue[0].value).toBe('1.5 hours');
      expect(tooltipValue[1].value).toBe('1.1 hours');
    });

    it('will apply a custom formatter when supplied', () => {
      const formatter = jest.fn();

      extractTimeSeriesTooltip(params, fakeChartTitle, formatter);
      expect(formatter).toHaveBeenCalledTimes(2);
    });
  });

  describe('secondsToDays', () => {
    const seconds = 151000;

    it('defaults to a single decimal', () => {
      expect(secondsToDays(seconds)).toBe('1.7');
    });

    it('will format to the specified precision', () => {
      expect(secondsToDays(seconds, 3)).toBe('1.748');
    });
  });

  describe('formatAsPercentage', () => {
    it('returns 0 if given NaN', () => {
      expect(formatAsPercentage(null)).toBe('0.0%');
      expect(formatAsPercentage('a')).toBe('0.0%');
    });

    it('formats valid values', () => {
      expect(formatAsPercentage(0.25)).toBe('25.0%');
      expect(formatAsPercentage('1.86', 0)).toBe('186%');
    });
  });

  describe('forecastDataToSeries', () => {
    let res;

    const forecastResponse = forecastDataToChartDate(
      deploymentFrequencyLastWeekData,
      mockLastWeekRawForecastData,
    );

    beforeEach(() => {
      res = forecastDataToSeries({
        forecastData: mockLastWeekRawForecastData,
        forecastHorizon: 3,
        forecastSeriesLabel: 'Forecast',
        dataSeries: mockLastWeekData,
        endDate: new Date('2015-07-04'),
      });
    });

    it('returns the data series to be displayed in charts', () => {
      expect(res).toEqual(forecastResponse);
    });

    it('includes the last data point from the data series', () => {
      expect(res[0]).toEqual(['Jul 03, 2015', 1]);
    });
  });

  describe('extractOverviewMetricsQueryParameters', () => {
    it('returns empty object when no requestParams property provided', () => {
      expect(extractOverviewMetricsQueryParameters()).toEqual({});
      expect(extractOverviewMetricsQueryParameters({})).toEqual({});
    });

    it.each`
      param           | value           | expected
      ${'start_date'} | ${'2024-01-01'} | ${'startDate'}
      ${'end_date'}   | ${'2024-12-31'} | ${'endDate'}
    `('correctly transforms the $param parameter', ({ param, value, expected }) => {
      const result = extractOverviewMetricsQueryParameters({ requestParams: { [param]: value } });
      expect(result[expected]).toBe(value);
    });
  });
});
