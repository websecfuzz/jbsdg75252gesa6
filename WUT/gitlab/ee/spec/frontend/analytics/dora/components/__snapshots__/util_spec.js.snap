// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ee/analytics/dora/components/util.js lead time data returns the correct lead time chart data after all processing of the API response 1`] = `
[
  {
    "areaStyle": {
      "color": "none",
    },
    "data": [
      [
        "Jun 27, 2015",
        86400,
      ],
      [
        "Jun 28, 2015",
        86400,
      ],
      [
        "Jun 29, 2015",
        86400,
      ],
      [
        "Jun 30, 2015",
        86400,
      ],
      [
        "Jul 01, 2015",
        432000,
      ],
      [
        "Jul 02, 2015",
        518400,
      ],
      [
        "Jul 03, 2015",
        604800,
      ],
    ],
    "itemStyle": {
      "color": "#a4a3a8",
    },
    "lineStyle": {
      "color": "#a4a3a8",
      "type": "dashed",
    },
    "name": "No data available",
    "showSymbol": false,
  },
  {
    "areaStyle": {
      "opacity": 0,
    },
    "data": [
      [
        "Jun 27, 2015",
        null,
      ],
      [
        "Jun 28, 2015",
        null,
      ],
      [
        "Jun 29, 2015",
        null,
      ],
      [
        "Jun 30, 2015",
        86400,
      ],
      [
        "Jul 01, 2015",
        432000,
      ],
      [
        "Jul 02, 2015",
        null,
      ],
      [
        "Jul 03, 2015",
        604800,
      ],
    ],
    "itemStyle": {
      "color": "#617ae2",
    },
    "lineStyle": {
      "color": "#617ae2",
    },
    "name": "Lead time",
    "showAllSymbol": true,
    "showSymbol": true,
    "symbolSize": 8,
  },
]
`;
